module/mempool/stdmap/backdata/arrayBackData.go

```diff
@@ -0,0 +1,400 @@
+	idPref, bucketIndex := a.idPrefixAndBucketIndex(entityId)
+	slotToUse, unique := a.slotInBucket(bucketIndex, idPref, entityId)
+	if !unique {
+		// entityId already exists
+		return false
+	}
+
+	if _, _, ok := a.linkedEntityOf(bucketIndex, slotToUse); ok {
+		// we are replacing an already linked (but old) key that has a valid value, hence
+		// we should remove its value from underlying entities list.
+		a.invalidateEntity(bucketIndex, slotToUse)
+	}
+
+	a.keyCount++
+	entityIndex := a.entities.Add(entityId, entity, a.ownerIndexOf(bucketIndex, slotToUse))
+	a.buckets[bucketIndex][slotToUse].keyIndex = a.keyCount
+	a.buckets[bucketIndex][slotToUse].entityIndex = entityIndex
+	a.buckets[bucketIndex][slotToUse].idPref = idPref
+	return true
+}
+
+// get retrieves the entity corresponding to given identifier from underlying entities list.
+// The boolean return value determines whether an entity with given id exists in the BackData.
+func (a *ArrayBackData) get(entityID flow.Identifier) (flow.Entity, bIndex, sIndex, bool) {
+	idPref, bucketIndex := a.idPrefixAndBucketIndex(entityID)
+	for slotIndex := sIndex(0); slotIndex < sIndex(bucketSize); slotIndex++ {
+		if a.buckets[bucketIndex][slotIndex].idPref != idPref {
+			continue
+		}
+
+		id, entity, linked := a.linkedEntityOf(bucketIndex, slotIndex)
+		if !linked {
+			// no linked entity for this (bucketIndex, slotIndex) pair.
+			return nil, 0, 0, false
+		}
+
+		if id != entityID {
+			// checking identifiers fully.
+			continue
+		}
+
+		return entity, bucketIndex, slotIndex, true
+	}
+
+	return nil, 0, 0, false
+}
+
+// idPrefixAndBucketIndex determines the id prefix as well as the bucket index corresponding to the
+// given identifier.
+func (a ArrayBackData) idPrefixAndBucketIndex(id flow.Identifier) (idPrefix, bIndex) {
+	// uint64(id[0:8]) used to compute bucket index for which this key (i.e., id) belongs to
+	bucketIndex := binary.LittleEndian.Uint64(id[0:8]) % a.bucketNum
+
+	// uint32(id[8:12]) used to compute a shorter key for this id to represent in memory.
+	idPref := binary.LittleEndian.Uint32(id[8:12])
+
+	return idPrefix(idPref), bIndex(bucketIndex)
+}
+
+// expiryThreshold returns the threshold for which all keys with index below threshold are considered old enough for eviction.
+func (a ArrayBackData) expiryThreshold() uint64 {
+	var expiryThreshold uint64 = 0
+	if a.keyCount > uint64(a.limit) {
+		// total number of keys written are above the predefined limit
+		expiryThreshold = a.keyCount - uint64(a.limit)
+	}
+
+	return expiryThreshold
+}
+
+// slotInBucket returns a free slot for this entityId in the bucket. In case the bucket is full, it invalidates the oldest (unlinked) key,
+// and returns its index as free slot. It returns false if the entityId already exists in this bucket.
+func (a *ArrayBackData) slotInBucket(bucketIndex bIndex, idPref idPrefix, entityId flow.Identifier) (sIndex, bool) {
```

Comment:
Hmm... instead of `bucketIndex bIndex` why not consider `b bucketIndex` ? Why? If `b` is only used in this function -- because it is a parameter -- then it's easy to look here and see that it's a `bucketIndex`. Also, the code below will get narrower and possibly easier to read? Also, why have a more description parameter instance than parameter type?

Reply:
`bucketIndex` is not limited to this single method solely. It is a general argument name across all of our methods. It is intentionally meant to be represented in this way to be clearly distinctive from the `slotIndex` when these two indices are co-located in the same method (e.g., `linkedEntityOf(bucketIndex bIndex, slotIndex sIndex)`). Having a clear naming distinction such as `bucketIndex` and `slotIndex` adds more readability, and also avoids confusions. For example in the same `linkedEntityOf` having `c.buckets[bucketIndex][slotIndex].entityIndex` is much more unambiguous and readable than `c.buckets[b][s].entityIndex`.

Reply:
> clean code 101: "If a name requires a comment, then the name does not reveal its intent."

I agree with that statement in general but I'm not sure it's super relevant here. Why? Because the `int` is not giving any hints away about what `d` is, and our case is not using a basic type like that with no hints. However, if the clean code 101 example was 'ported' to our scenario then `int d` would be `elapsedTimeInDays d;` (a 3rd possible example) which has no comment and is arguably more readable later on... BUT common wisdom says it's only more readable later on IF the declaration is close (in source lines) to the use of `d`. If the reader has to scroll around, search, or look in another file etc to discover what `d` is then that's not more readable...

So I think the clean code 101 suggestion should be changed to "If a local declaration requires a comment, then the name and/or type does not reveal its intent."

> having `c.buckets[bucketIndex][slotIndex].entityIndex` is much more unambiguous and readable than `c.buckets[b][s].entityIndex`

I agree for this example. BUT... I would also suggest changing the structs around to make this more intention revealing in general and regardless of whether using the shorter `s` for `slot` in the end, or not :-) Why?

`c.buckets[bucketIndex][slotIndex].entityIndex` <-- the `[][]` looks and is more intimidating to the reader

`c.buckets[bucketIndex].slots[slotIndex].entityIndex` <-- looks more intention revealing to the reader

`c.buckets[b].slots[s].entityIndex` <-- the best? intention revealing and short?

In the last example then it's obvious that `b` is for `buckets` and `s` is for `slots`, or?

Reply:
https://github.com/onflow/flow-go/pull/1778/commits/ec671d46fa8f8348f73e2e636640d54c720350fd